### PR TITLE
fix: use edge version of loki-k8s to stabilise integration tests

### PR DESCRIPTION
**Pull request by yanksyoon**
The latest/stable version of loki-k8s charm isn't quite stable due to the missing `can_connect` guard (please refer to https://chat.charmhub.io/charmhub/pl/xgqyman4btym3ff74fcjomzxbw).

This PR proposes to use the latest revision (82) of the loki-k8s charm to resolve the issues with flaky tests.

Other minor improvements include:
- Updating pre-run script to remove duplicate `microk8s enable registry` (already enabled on operator-workflows)
- Updating Required Status Checks action
- Updating libraries
- Updating typing syntax to match Python 3.8
- Updating libraries

### CODE REVIEW

**Review by arturo-seijas — DISMISSED** [review]
LGTM

**Review by arturo-seijas — DISMISSED** [review]

File: tests/integration/conftest.py [hunk 1]
```diff
@@ -449,7 +440,7 @@ async def prometheus_fixture(
     model: Model, application_name: str
 ) -> typing.AsyncGenerator[Application, None]:
     """Deploy and yield prometheus charm application with relation to WordPress charm."""
-    prometheus = await model.deploy("prometheus-k8s", channel="stable", trust=True)
+    prometheus = await model.deploy("prometheus-k8s", channel="latest/edge", trust=True)
```
Comment (yhaliaw) [inline tests/integration/conftest.py:443]: Is this change to `latest/edge` temporary, like for Loki?
If so it might be good to add a comment, just like for Loki.

File: tests/integration/conftest.py [hunk 2]
```diff
@@ -475,7 +469,7 @@ async def grafana_fixture(
     model: Model, application_name: str
 ) -> typing.AsyncGenerator[Application, None]:
     """Deploy and return grafana charm application with relation to WordPress charm."""
-    grafana = await model.deploy("grafana-k8s", channel="stable", trust=True)
+    grafana = await model.deploy("grafana-k8s", channel="latest/edge", trust=True)
```
Comment (yhaliaw) [inline tests/integration/conftest.py:472]: Same as above.

File: tests/integration/test_core.py [hunk 3]
```diff
@@ -576,7 +577,29 @@ async def test_loki_integration(
     assert: loki joins relation successfully, logs are being output to container and to files for
         loki to scrape.
     """
-    await model.wait_for_idle(apps=[application_name, loki.name], status="active")
+
+    async def wait_loki_unit_agent_idle():
```
Comment (amandahla) [inline tests/integration/test_core.py:581]: I'm curious: Why is this needed?

Comment (yanksyoon) [reply tests/integration/test_core.py:581]: The agent can still be executing when the `model.wait_for_idle` returns (this function isn't quite 100%), and when the assertions (querying from the loki api) are executed, it returns the wrong result which doesn't contain the access logs and error logs. It's a messy workaround for having no `agent.wait_for_idle()` or `unit.wait_for_idle().
I've tested the version with and without on the actions (the tests always passes on my local ;_;), the pass rate with this workaround is much higher than the ones without. :)
Thanks for the review!

File: tests/integration/conftest.py (hunk 1 above)
Comment (yanksyoon) [reply tests/integration/conftest.py:445]: Added comments for all the `edge` deployments. Thanks for the suggestion!

File: tests/integration/conftest.py (hunk 2 above)
Comment (yanksyoon) [reply tests/integration/conftest.py:476]: Added comments for all the `edge` deployments. Thanks for the suggestion!

File: tests/integration/test_core.py (hunk 3 above)
Comment (amandahla) [reply tests/integration/test_core.py:581]: It's a nice workaround :) What do you think about adding a small comment regarding that? Usually, we see only the model.wait_for_idle.

Comment (yanksyoon) [reply tests/integration/test_core.py]: Refactored to a helper with the description in the docstring! Thanks for the suggestion :)

**Review by weiiwang01 — APPROVED** [review]

**Review by nrobinaubertin — APPROVED** [review]

**Review by arturo-seijas — APPROVED** [review]
